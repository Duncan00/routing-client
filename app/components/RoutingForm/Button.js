import styled from 'styled-components';

const Button = styled.button`
  width: 100px;
`;

export default Button;
