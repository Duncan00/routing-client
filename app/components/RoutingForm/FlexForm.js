import styled from 'styled-components';
import { Form } from 'formik';

const FlexForm = styled(Form)`
  padding: 30px 15px;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  max-height: 360px;

  @media only screen and (max-width: 600px) {
    width: 50%;
  }

  @media only screen and (min-width: 600px) {
    width: 360px;
  }
`;

export default FlexForm;
