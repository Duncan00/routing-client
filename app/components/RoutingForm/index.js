import React from 'react';
import PropTypes from 'prop-types';
import { Formik } from 'formik';
import * as Yup from 'yup';
import Error from '../Error';
import Group from '../Group';
import FlexForm from './FlexForm';
import ButtonGroup from './ButtonGroup';
import Button from './Button';
import FormInput from '../FormInput';

const RoutingForm = ({ route, onSubmitForm, onResetForm, error, loading }) => {
  const { total_distance: totalDistance, total_time: totalTime } = route;

  return (
    <Formik
      initialValues={{
        starting_location: '',
        drop_off_point: '',
      }}
      validationSchema={Yup.object({
        starting_location: Yup.string().required('Required!'),
        drop_off_point: Yup.string().required('Required!'),
      })}
      onSubmit={onSubmitForm}
      onReset={onResetForm}
    >
      {({ isSubmitting, setFieldValue, values }) => (
        <FlexForm>
          <FormInput
            name="starting_location"
            displayName="Starting Location"
            value={values.starting_location}
            setFieldValue={setFieldValue}
          />

          <FormInput
            name="drop_off_point"
            displayName="Drop Off Point"
            value={values.drop_off_point}
            setFieldValue={setFieldValue}
          />

          <Group />

          <Group>
            {totalDistance && <label>total distance: {totalDistance}</label>}
            {totalTime && <label>total time: {totalTime}</label>}
          </Group>

          {error && <Error>{error}</Error>}

          <ButtonGroup>
            <Button type="submit" disabled={isSubmitting || loading}>
              {getSubmitButtonWord({ isSubmitting, loading, route, error })}
            </Button>
            <Button type="reset" disabled={isSubmitting || loading}>
              Reset
            </Button>
          </ButtonGroup>
        </FlexForm>
      )}
    </Formik>
  );
};

export function getSubmitButtonWord({ isSubmitting, loading, route, error }) {
  if (isSubmitting || loading) {
    return 'Loading...';
  }
  if (route || error) {
    return 'Re-Submit';
  }
  return 'Submit';
}

RoutingForm.propTypes = {
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,
  loading: PropTypes.bool.isRequired,
  onSubmitForm: PropTypes.func.isRequired,
  onResetForm: PropTypes.func.isRequired,
  route: PropTypes.oneOfType([PropTypes.object, PropTypes.bool]),
};

export default RoutingForm;
