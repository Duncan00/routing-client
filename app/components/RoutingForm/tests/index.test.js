import React from 'react';
import { render } from 'react-testing-library';
import RoutingForm, { getSubmitButtonWord } from '../index';
import setGlobalGoogleMock from '../../../../internals/testing/setGlobalGoogleMock';

describe('<RoutingForm />', () => {
  it('should render RoutingForm', () => {
    setGlobalGoogleMock();

    const { container } = render(
      <RoutingForm
        loading
        error
        onResetForm={() => {}}
        onSubmitForm={() => {}}
        route={{ total_distance: 1000, total_time: 100 }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('getSubmitButtonWord', () => {
  describe('isSubmitting === true', () => {
    it('should return Loading...', () => {
      const buttonWord = getSubmitButtonWord({
        isSubmitting: true,
        loading: false,
        route: false,
        error: false,
      });
      expect(buttonWord).toEqual('Loading...');
    });
  });

  describe('loading === true', () => {
    it('should return Loading...', () => {
      const buttonWord = getSubmitButtonWord({
        isSubmitting: false,
        loading: true,
        route: false,
        error: false,
      });
      expect(buttonWord).toEqual('Loading...');
    });
  });

  describe('route exists', () => {
    it('should return Re-Submit', () => {
      const buttonWord = getSubmitButtonWord({
        isSubmitting: false,
        loading: false,
        route: {},
        error: false,
      });
      expect(buttonWord).toEqual('Re-Submit');
    });
  });

  describe('error exists', () => {
    it('should return Re-Submit', () => {
      const buttonWord = getSubmitButtonWord({
        isSubmitting: false,
        loading: false,
        route: false,
        error: {},
      });
      expect(buttonWord).toEqual('Re-Submit');
    });
  });

  describe('In default case', () => {
    it('should return Submit', () => {
      const buttonWord = getSubmitButtonWord({
        isSubmitting: false,
        loading: false,
        route: false,
        error: false,
      });
      expect(buttonWord).toEqual('Submit');
    });
  });
});
