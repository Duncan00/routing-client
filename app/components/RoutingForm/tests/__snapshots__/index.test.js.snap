// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoutingForm /> should render RoutingForm 1`] = `
.c5 {
  color: red;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

.c0 {
  padding: 30px 15px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  max-height: 360px;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c7 {
  width: 100px;
}

.c4 {
  padding: 0;
  border: none;
  background: none;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c3 {
  width: 90%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media only screen and (max-width:600px) {
  .c0 {
    width: 50%;
  }
}

@media only screen and (min-width:600px) {
  .c0 {
    width: 360px;
  }
}

<div>
  <form
    action="#"
    class="c0"
  >
    <div
      class="c1"
    >
      <label
        for="starting_location"
      >
        Starting Location
      </label>
      <div
        class="c2"
      >
        <input
          class="c3"
          name="starting_location"
          value="Innocentre, Hong Kong"
        />
        <button
          class="c4"
          type="button"
        >
          ✖
        </button>
      </div>
      <div
        class="c5"
      />
    </div>
    <div
      class="c1"
    >
      <label
        for="drop_off_point"
      >
        Drop Off Point
      </label>
      <div
        class="c2"
      >
        <input
          class="c3"
          name="drop_off_point"
          value="Hong Kong International Airport Terminal 1"
        />
        <button
          class="c4"
          type="button"
        >
          ✖
        </button>
      </div>
      <div
        class="c5"
      />
    </div>
    <div
      class="c1"
    />
    <div
      class="c1"
    >
      <label>
        total distance: 
        1000
      </label>
      <label>
        total time: 
        100
      </label>
    </div>
    <div
      class="c5"
    />
    <div
      class="c6"
    >
      <button
        class="c7"
        disabled=""
        type="submit"
      >
        Loading...
      </button>
      <button
        class="c7"
        disabled=""
        type="reset"
      >
        Reset
      </button>
    </div>
  </form>
</div>
`;
